tests/unit/test_charm.py

```diff
@@ -0,0 +1,294 @@
+
+from ops import testing
+from ops.model import ActiveStatus, ModelError, WaitingStatus
+from ops.pebble import ExecError
+
+from charm import UPFOperatorCharm
+
+
+class TestCharm(unittest.TestCase):
+    @patch("lightkube.core.client.GenericSyncClient")
+    @patch(
+        "charm.KubernetesServicePatch",
+        lambda charm, ports: None,
+    )
+    def setUp(self, patch_k8s_client):
+        self.namespace = "whatever"
+        self.harness = testing.Harness(UPFOperatorCharm)
+        self.harness.set_model_name(name=self.namespace)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    @patch("ops.model.Container.exec", new=Mock)
+    @patch("ops.model.Container.exists")
+    @patch("ops.model.Container.push")
+    def test_given_bessd_config_file_not_written_when_bessd_pebble_ready_then_config_file_is_written(  # noqa: E501
+        self,
+        patch_push,
+        patch_exists,
+    ):
+        patch_exists.return_value = False
+
+        self.harness.container_pebble_ready(container_name="bessd")
+
+        patch_push.assert_has_calls(
+            [
+                call(
+                    path="/etc/bess/conf/upf.json",
+                    source='{\n  "access": {\n    "ifname": "access"\n  },\n  "core": {\n    "ifname": "core"\n  },\n  "cpiface": {\n    "dnn": "internet",\n    "enable_ue_ip_alloc": false,\n    "hostname": "sdcore-upf.whatever.svc.cluster.local",\n    "http_port": "8080",\n    "ue_ip_pool": "172.250.0.0/16"\n  },\n  "enable_notify_bess": true,\n  "gtppsc": true,\n  "hwcksum": true,\n  "log_level": "trace",\n  "max_sessions": 50000,\n  "measure_flow": false,\n  "measure_upf": true,\n  "mode": "af_packet",\n  "notify_sockaddr": "/pod-share/notifycp",\n  "qci_qos_config": [\n    {\n      "burst_duration_ms": 10,\n      "cbs": 50000,\n      "ebs": 50000,\n      "pbs": 50000,\n      "priority": 7,\n      "qci": 0\n    }\n  ],\n  "slice_rate_limit_config": {\n    "n3_bps": 1000000000,\n    "n3_burst_bytes": 12500000,\n    "n6_bps": 1000000000,\n    "n6_burst_bytes": 12500000\n  },\n  "table_sizes": {\n    "appQERLookup": 200000,\n    "farLookup": 150000,\n    "pdrLookup": 50000,\n    "sessionQERLookup": 100000\n  },\n  "workers": 1\n}',  # noqa: E501
+                ),
+            ]
+        )
+
+    @patch("ops.model.Container.exec", new=Mock)
+    @patch("ops.model.Container.exists")
+    @patch("ops.model.Container.push")
+    def test_given_bessd_config_file_written_when_bessd_pebble_ready_then_config_file_is_not_written(  # noqa: E501
+        self,
+        patch_push,
+        patch_exists,
+    ):
+        patch_exists.return_value = True
+
+        self.harness.container_pebble_ready(container_name="bessd")
+
+        patch_push.assert_not_called()
+
+    @patch("ops.model.Container.exec", new=Mock())
+    @patch("ops.model.Container.exists")
+    def test_given_bessd_config_file_is_written_when_bessd_pebble_ready_then_pebble_plan_is_applied(  # noqa: E501
```

Comment:
Are we testing the plan is applied, or that is the one we expect? The plan could have been applied and be the same, it clould have been already applied and be correct, and it cloud have been applied, changed, and correct. IDK if I'm making my point, probably is just a matter of naming of the test.

Reply:
Added "expected" in the test name, hopefully it's clearer.